Implement tier-2 JIT compiler to accelerate ISS

qwe661234 opened the pull request:
The T2C, designed for flexibility, efficiently transforms custom IR within a chained block into LLVM IR. After the translation, the built LLVM IR is offloaded to the LLVM backend,  where it undergoes optimization through several selected LLVM passes. Subsequently, the optimized LLVM IR is passed to the LLVM execution engine, which compiles the optimized LLVM IR and returns a function pointer to the generated machine code. Because our design for T2C opts for a function pointer of the generated machine code instead of storing it in the code cache, we need not account for the binary format discrepancies among different compilers. This process allows for the possibility of substituting LLVM with alternative optimization frameworks if required.

Figure below shows the performance analysis of QEMU and rv32emu tiered-JIT mode.
![image](https://github.com/sysprog21/rv32emu/assets/48278026/3022573e-b525-4b02-936b-4de2798574a7)

jserv reviewed (commented):
<!-- github-benchmark-action-comment(start): Benchmarks Summary -->
# Benchmarks

<details>

| Benchmark suite | Current: 5fb1609a6ee34dd89769887d10bd1ae6077a65ce | Previous: 6c13a845d4ef1f6cadd5d1592e8b029e1f5ec089 | Ratio |
|-|-|-|-|
| `Dhrystone` | `3.44` Average DMIPS over 10 runs | `5.87` Average DMIPS over 10 runs | `1.71` |
| `Coremark` | `0.004` Average iterations/sec over 10 runs | `0.004` Average iterations/sec over 10 runs | `1` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).
<!-- github-benchmark-action-comment(end): Benchmarks Summary -->

jserv reviewed (changes requested):
Rebase the latest `master` branch.

jserv commented on .github/workflows/main.yml:
```diff
@@ -65,14 +66,16 @@ jobs:
-            make clean && make ENABLE_EXT_A=0 ENABLE_JIT=1 check -j$(nproc)
-            make clean && make ENABLE_EXT_F=0 ENABLE_JIT=1 check -j$(nproc)
-            make clean && make ENABLE_EXT_C=0 ENABLE_JIT=1 check -j$(nproc)
+            export VER=-14
```
Why `VER` here?

qwe661234 replied:
Because the cloud machine can not find the `llvm-config` when running GitHub CI, but `llvm-config-14` can be found.

jserv replied:
Can you support [recent LLVM releases](https://releases.llvm.org/)? e.g., LLVM 18.x

jserv commented on Makefile:
```diff
@@ -125,14 +125,20 @@ endif
 $(call set-feature, JIT)
 ifeq ($(call has, JIT), 1)
-OBJS_EXT += jit.o
+OBJS_EXT += jit.o t2jit.o
```
All tired-2 JIT compiler (T2C) related files should have the prefix `t2c`. That is, `t2c.h`, `t2c.c`, and `t2c_template.c`.

jserv commented on Makefile:
```diff
@@ -125,14 +125,20 @@ endif
 ifeq ($(call has, JIT), 1)
-OBJS_EXT += jit.o
+OBJS_EXT += jit.o t2jit.o
+CFLAGS += -g `llvm-config${VER} --cflags`
```
T2C is only built when the supported versions of LLVM packages are available. That is, T2C is ignored if the LLVM check cannot be satisfied.

jserv commented on src/t2_rv32_template.c:
```diff
@@ -0,0 +1,2143 @@
+
```
Add comments at the beginning of this file.

jserv commented on src/t2_rv32_template.c:
```diff
@@ -0,0 +1,2143 @@
+RVT2OP(nop, { return; })
+
+RVT2OP(lui, {
+    LLVMValueRef rd_offset[1] = {LLVMConstInt(
```
Could you provide some helper macros to simplify the invocation of LLVM APIs? This could reduce the effort required when we need to upgrade LLVM versions.

jserv commented on src/t2jit.c:
```diff
@@ -0,0 +1,222 @@
+#include <llvm-c/Analysis.h>
```
Add copyright notice.

jserv commented on src/emulate.c:
```diff
@@ -1143,9 +1146,23 @@ void rv_step(void *arg)
+            prev = NULL;
+            continue;
+        } /* check if invoking times of t1 generated code exceed threshold */
+        else if (block->n_invoke >= THRESHOLD) {
```
Replace `else if` with `if` statement.

jserv commented:
LLVM-17 is the preferable version for T2C's LLVM integration. Check [Linux_Build.yml](https://github.com/wheremyfoodat/Panda3DS/blob/master/.github/workflows/Linux_Build.yml) for the usage. Alternatively, [xPack built LLVM](https://xpack.github.io/dev-tools/clang/) packages are available for several host environments.

jserv commented on Makefile:
```diff
@@ -125,7 +125,16 @@ endif
+ifneq ("$(shell which llvm-config-17)", "")
+ENABLE_T2C := 1
+OBJS_EXT += t2c.o
+CFLAGS += -g `llvm-config-17 --cflags`
```
Replace with `$(shell llvm-config-17 --cflags)`.

jserv commented on src/t2c.c:
```diff
@@ -0,0 +1,227 @@
+        code;                                                             \
+    }
+
+#include "t2c_rv32_template.c"
```
Rename to `t2c_template.c`

jserv commented on src/t2c.h:
```diff
@@ -0,0 +1,10 @@
+
+#include "riscv_private.h"
+
+typedef intptr_t (*funcPtr_t)(riscv_t *);
```
Enforce the consistent style as mentioned in `CONTRIBUTING.md`.
Of course, all LLVM related APIs are exceptional.

jserv commented on .github/workflows/main.yml:
```diff
@@ -43,6 +43,9 @@ jobs:
             sudo apt-get install -q -y libsdl2-dev libsdl2-mixer-dev
+            wget https://apt.llvm.org/llvm.sh
+            sudo chmod +x ./llvm.sh
+            sudo ./llvm.sh 17
```
Inside the guest environment initiated by [run-on-arch-action](https://github.com/uraimo/run-on-arch-action), the command `sudo` is not allowed. Instead, superuser (aka "root") is the only user there.

jserv commented on src/t2c_rv32_template.c:
```diff
@@ -0,0 +1,1451 @@
+ * generated machine code.
+ */
+
+FORCE_INLINE LLVMValueRef build_rs1_addr(LLVMValueRef start,
```
Does it make more sense to move these helper functions out of `t2c_template.c`?

jserv commented on Makefile:
```diff
@@ -125,7 +125,16 @@ endif
 ifeq ($(call has, JIT), 1)
-OBJS_EXT += jit.o
+OBJS_EXT += jit.o 
+ifeq ("$(shell llvm-config-17 --cflags)", "-I/usr/lib/llvm-17/include  -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS")
```
This check might not be invalid for different Linux distributions and macOS (via homebrew).

jserv commented:
Consider the following quick and dirty fix for macOS enablement:
```diff
--- a/Makefile
+++ b/Makefile
@@ -125,16 +125,28 @@ endif
 ENABLE_JIT ?= 0
 $(call set-feature, JIT)
 ifeq ($(call has, JIT), 1)
-OBJS_EXT += jit.o 
-ifneq ("$(findstring -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS, "$(shell llvm-config-17 --cflags)")", "")
+OBJS_EXT += jit.o
+LLVM_CONFIG = llvm-config-17
+LLVM_CONFIG := $(shell which $(LLVM_CONFIG))
+ifndef LLVM_CONFIG
+    # Try Homebrew on macOS
+    LLVM_CONFIG = /opt/homebrew/opt/llvm@17/bin/llvm-config
+    LLVM_CONFIG := $(shell which $(LLVM_CONFIG))
+    ifdef LLVM_CONFIG
+    LDFLAGS += -L/opt/homebrew/opt/llvm@17/lib
+    endif
+endif
+ifneq ("$(LLVM_CONFIG)", "")
+ifneq ("$(findstring -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS, "$(shell $(LLVM_CONFIG) --cflags)")", "")
 ENABLE_T2C := 1
 OBJS_EXT += t2c.o
-CFLAGS += -g `llvm-config-17 --cflags`
-LDFLAGS += `llvm-config-17 --libs core executionengine mcjit interpreter analysis native --system-libs`
+CFLAGS += -g $(shell $(LLVM_CONFIG) --cflags)
+LDFLAGS += $(shell $(LLVM_CONFIG) --libs core executionengine mcjit interpreter analysis native --system-libs)
 else
 ENABLE_T2C := 0
 $(warning No llvm-config-17 installed. Check llvm-config-17 installation in advance)
 endif
+endif
 ifneq ($(processor),$(filter $(processor),x86_64 aarch64 arm64))
 $(error JIT mode only supports for x64 and arm64 target currently.)
 endif
```

jserv reviewed (changes requested):
Update the top-level documentation for the dependent packages for tier-2 JIT compiler.
* Debian/Ubuntu: `apt-get install llvm-17`
* macOS: `brew install llvm@17`

jserv commented on Makefile:
```diff
@@ -125,7 +125,16 @@ endif
-OBJS_EXT += jit.o
+OBJS_EXT += jit.o 
+ifneq ("$(findstring -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS, "$(shell llvm-config-17 --cflags)")", "")
+ENABLE_T2C := 1
```
Add some comments here. e.g., tier-2 JIT compiler powered by LLVM.

jserv commented on src/emulate.c:
```diff
@@ -32,6 +32,9 @@ extern struct target_ops gdbstub_ops;
 #include "cache.h"
 #include "jit.h"
+#if RV32_HAS(T2C)
+#include "t2c.h"
```
The header `t2c.h` should be included in `jit.h`.

jserv commented on src/emulate.c:
```diff
@@ -1137,15 +1143,31 @@ void rv_step(void *arg)
         last_pc = rv->PC;
 #if RV32_HAS(JIT)
+#if RV32_HAS(T2C)
+        /* execute by tier-2 JIT compiler */
```
Improve the grammar.

jserv commented on src/t2c.h:
```diff
@@ -0,0 +1,8 @@
+/*
```
The header `t2c.h` might not be necessary. Merge it into `jit.h`.

jserv commented on Makefile:
```diff
@@ -125,7 +125,32 @@ endif
+  # Try Homebrew on macOS
+  LLVM_CONFIG = /opt/homebrew/opt/llvm@17/bin/llvm-config
+  LLVM_CONFIG := $(shell which $(LLVM_CONFIG))
+ifdef LLVM_CONFIG
```
Indent properly.

jserv commented on README.md:
```diff
@@ -30,7 +30,7 @@ Features:
 * Experimental SDL-based display/event/audio system calls for running video games
 * Support for remote GDB debugging
-* Experimental JIT compiler for performance boost while maintaining a small footprint
+* Two-tier JIT compiler for performance boost while maintaining a small footprint
```
Replace with "Tiered JIT compilation"

jserv commented on README.md:
```diff
@@ -40,6 +40,12 @@ and [SDL2_Mixer library](https://wiki.libsdl.org/SDL2_mixer) installed.
 * Ubuntu Linux / Debian: `sudo apt install libsdl2-dev libsdl2-mixer-dev`
 
+### JIT compiler
+The tier-2 JIT compiler in `rv32emu` leverages LLVM for powerful optimization. Therefore, the target system must have [`LLVM`](https://github.com/llvm/llvm-project) installed, with version 17 recommended. If `LLVM` is not installed, only the tier-1 JIT compiler will be used for performance enhancement.
```
Specify LLVM website: https://llvm.org/

jserv commented on README.md:
```diff
@@ -81,7 +87,7 @@ The image containing all the necessary tools for development and testing can be
 * `ENABLE_GDBSTUB` : GDB remote debugging support
 * `ENABLE_SDL` : Experimental Display and Event System Calls
-* `ENABLE_JIT` : Experimental JIT compiler
+* `ENABLE_JIT` : Two-tier JIT compiler
```
Don't change this.

jserv commented on src/t2c.c:
```diff
@@ -0,0 +1,314 @@
+                                        LLVMBasicBlockRef *entry,
+                                        uint32_t pc)
+{
+    struct LLVM_block_map_entry map_entry;
```
Use C99 designated initializers when possible.

jserv commented on src/t2c.c:
```diff
@@ -0,0 +1,314 @@
+        LLVMInt32Type(), offsetof(riscv_t, X) / sizeof(int) + ir->rs1, true)};
+    return LLVMBuildInBoundsGEP2(*builder, LLVMInt32Type(),
+                                 LLVMGetParam(start, 0), rs1_offset, 1,
+                                 "addr_rs1");
```
C language standard (i.e. C99 and C11) defines a predefined identifier as follows in clause 6.4.2.2:
> "The identifier `__func__` shall be implicitly declared by the translator as if, immediately following the opening brace of each function definition, the declaration
```c
static const char __func__[] = "function-name";
```
> appeared, where function-name is the name of the lexically-enclosing function."

It means that the C compiler implicitly adds `__func__` in every function so that it can be used in that function to get the function name.

jserv commented:
I am a bit worried about the lengthy file src/t2c_template.c, which could lead to potential maintenance issues. Can you suggest a more advanced way to reduce the code size for T2C?

qwe661234 commented:
> I am a bit worried about the lengthy file src/t2c_template.c, which could lead to potential maintenance issues. Can you suggest a more advanced way to reduce the code size for T2C?

I reduce the code size through inline function and MARCO, this modification also provides more information about the gnerated machine code and make it more readable.

jserv reviewed (changes requested):
Rebase the latest `master` branch since libc substitution is dropped.

jserv reviewed (changes requested):
For consistent naming scheme, enforce the following rules:
- functions: specify prefix `t2c_`. e.g., `t2c_compile`
- macros: specify prefix `T2C`. e.g., `T2C_OP` and `T2C_LLVM_STORE_IMM32`

jserv commented on src/t2c.c:
```diff
@@ -0,0 +1,302 @@
+        code;                                                             \
+    }
+
+#define T2C_BUILD_LLVM_ADDR(reg, rv_member, ir_member)                       \
```
Should be `T2C_LLVM_BUILD_ADDR`. It can even be simpler as `T2C_LLVM_GEN`.

jserv commented on src/t2c.c:
```diff
@@ -0,0 +1,302 @@
+T2C_BUILD_LLVM_ADDR(sp, X, rv_reg_sp);
+T2C_BUILD_LLVM_ADDR(PC, PC, 0);
+
+#define T2C_BUILD_LLVM_STORE_IMM32(builder, val, addr) \
```
The prefix `T2C_` should be applied to the implementation details, and the macros related to LLVM should have prefix `T2C_LLVM_`.

jserv commented:
I defer to @vacantron for reviewing.

vacantron commented on src/t2c.c:
```diff
@@ -0,0 +1,302 @@
+            LLVMInt32Type(),                                                 \
+            offsetof(riscv_t, rv_member) / sizeof(int) + ir_member, true)};  \
+        return LLVMBuildInBoundsGEP2(*builder, LLVMInt32Type(),              \
+                                     LLVMGetParam(start, 0), offset, 1, ""); \
```
Is this statement better?

```c
LLVMValueRef offset = LLVMConstInt(LLVMInt32Type(), offsetof(riscv_t, rv_member) / sizeof(int) + ir_member, true);
return LLVMBuildInBoundsGEP2(*builder, LLVMInt32Type(), LLVMGetParam(start, 0), &offset, 1, "");
```

vacantron commented on src/t2c.c:
```diff
@@ -0,0 +1,302 @@
+    LLVMValueRef io_param[1] = {LLVMGetParam(start, 0)};
+    LLVMBuildCall2(*builder,
+                   LLVMFunctionType(LLVMVoidType(), param_types, 1, 0), io_func,
+                   io_param, 1, "");
```
Ditto.

vacantron commented on src/t2c_template.c:
```diff
@@ -0,0 +1,847 @@
+            true)};
+        LLVMValueRef addr_rd = LLVMBuildInBoundsGEP2(*builder, LLVMInt32Type(),
+                                                     LLVMGetParam(start, 0),
+                                                     rd_offset, 1, "addr_rd");
```
Ditto.

vacantron commented on src/t2c_template.c:
```diff
@@ -0,0 +1,847 @@
+T2C_OP(addi, {
+    T2C_LLVM_GEN_LOAD_VMREG(rs1, 32, t2c_gen_rs1_addr(start, builder, ir));
+    LLVMValueRef res = T2C_LLVM_GEN_ALU32_IMM(Add, val_rs1, ir->imm);
+    LLVMBuildStore(*builder, res, t2c_gen_rd_addr(start, builder, ir));
```
Can we avoid the storing of result after computing here or would it be optimized out in LLVM Pass?

qwe661234 replied:
Yes, it will be optimized through LLVM Pass. To be specific, it will be optimized during register allocation stage of LLVM backend.

vacantron commented on src/t2c.c:
```diff
@@ -0,0 +1,302 @@
+
+#define T2C_OP(inst, code)                                                \
+    static void t2c_##inst(                                               \
+        LLVMBuilderRef *builder UNUSED, LLVMTypeRef *param_types UNUSED,  \
```
The `LLVMBuilderRef *builder` is for receiving the change of control flow in most `slt`-like RISC-V insturctions.
Can we use `LLVMBuildSelect` alternatively to eliminate the pinter to builder and reduce the potential control flow change?

qwe661234 replied:
I have tried `LLVMBuildSelect` before the current implementation, but it would cause error. For example, you can try to replace it with `LLVMBuildSelect(*builder, cmp, T2C_LLVM_GEN_STORE_IMM32(*builder, 1, addr_rd), T2C_LLVM_GEN_STORE_IMM32(*builder, 0, addr_rd),"");`, and make the simulation exeute only through T2C path.

qwe661234 replied:
The error output of running `fcalc.elf` is
```
./build/rv32emu ./build/fcalc.elf 
Test 'Integer' failed; Expected 0.00000000000000000000000000000000000000000011/0, got 0.000000000000000000000000000000000000000000141937561/0
11Integer11.32Real number.32Sub-0-Real-42Negation1.2 + 5Addition3 - 1.4Subtraction10 * 5Multiplication50 / 10Division3 - 2 * 5PrecedencepiPi3 * (1 + 2)Parentheses(1 + 2Unmatched (Performed 0 tests, 0 failures, 0% success rate.
INFinfNANnan0123456789abcdef0123456789ABCDEF(null)0/POSIXCASCIIISO-8859-1cjksinglecjknarrowcjkwideUTF-8UTF8JISEUCJPEUCJPSJISISO8859ISO-8859-CPKOI8CP20866CP21866GEORGIANPSCP101PT154CP102TIS620CP874LC_ALLLANG.LC_COLLATELC_CTYPELC_MONETARYLC_NUMERICLC_TIMELC_MESSAGESInfinityNaNBalloc succeeded/Users/ilg/actions-runners/xpack-dev-tools/_work/riscv-none-elf-gcc-xpack/riscv-none-elf-gcc-xpack/build/darwin-arm64/sources/newlib-4.3.0.20230120/newlib/libc/stdlib/dtoa.c/Users/ilg/actions-runners/xpack-dev-tools/_work/riscv-none-elf-gcc-xpack/riscv-none-elf-gcc-xpack/build/darwin-arm64/sources/newlib-4.3.0.20230120/newlib/libc/stdlib/mprec.c, function: assertion "(null)" failed: file "(null)", line 011(null)
$��,$�������$�������4�������$���$���$���4���$���������������4���
��,
����,
��,
��,
���
   ��,
��,
����$��,
�������,
������  ���     ���     ���     ���     ���     ���     ���     ���     ��,
��,
��,
��,
��,
��,
��,
��,
��,\��h
��,
��,
��,
��P��,
��,
�����,
��,
��,
��`��,
��4��,
��,
��P��,
��,
��,
��,
��,
��,
��,
��,
��,
����������,
��<��,
��h�����(�����,
��`��8
      �����,
��,
```
Therefore, I believe the current implementation is feasible, as the use of `LLVMBuildSelect` results in incorrect behavior.

vacantron replied:
I tried this and it works for me. It seems that `LLVMBuilderSelect` needs immediate value to be its operands.

```c
LLVMValueRef r1 = LLVMConstInt(LLVMInt32Type(), 1, true);
LLVMValueRef r2 = LLVMConstInt(LLVMInt32Type(), 0, true);
LLVMValueRef _ = LLVMBuildSelect(*builder, cmp, r1, r2, "");
LLVMBuildStore(*builder, _, addr_rd);
```

qwe661234 replied:
Right, it works for me as well. I misunderstood the use of `LLVMBuildSelect`. I have now modified the implementation accordingly.

qwe661234 committed:
Implement tier-2 JIT compiler to accelerate ISS

The T2C, designed for flexibility, efficiently transforms custom IR
within a chained block into LLVM IR. After the translation, the built
LLVM IR is offloaded to the LLVM backend, where it undergoes
optimization through several selected LLVM passes. Subsequently, the
optimized LLVM IR is passed to the LLVM execution engine, which compiles
the optimized LLVM IR and returns a function pointer to the generated
machine code. Because our design for T2C opts for a function pointer of
the generated machine code instead of storing it in the code cache, we
need not account for the binary format discrepancies among different
compilers. This process allows for the possibility of substituting LLVM
with alternative optimization frameworks if required.

henrybear327 commented on .github/workflows/main.yml:
```diff
@@ -65,14 +68,14 @@ jobs:
       run: |
             make clean && make ENABLE_UBSAN=1 check -j$(nproc)
-            make clean && make ENABLE_JIT=1 ENABLE_UBSAN=1 check -j$(nproc)
+            make ENABLE_JIT=1 clean clean && make ENABLE_JIT=1 ENABLE_UBSAN=1 check -j$(nproc)
```
Out of curiosity, is there a reason to run the `clean` target twice? @qwe661234

qwe661234 replied:
Sorry, it is my mistake.
